frontend/src/pedigree.tsx

```diff
@@ -1,73 +1,165 @@
+               : ind.sex == 'female' ? 'oval'
+                                     : 'triangle', // unknown sex
+          color: color
+        }
+}
 
-export function PedigreeNetwork({ pedigree }: { pedigree: { edges: any[], nodes: any[]} }){
-  // A reference to the div rendered by this component
-  const domNode = useRef(null);
-  const network = useRef(null);
-  const history = useHistory();
-
-  const options = {
-    width: Math.round(window.innerWidth * 0.85) + 'px',//width is calculated considering the parent component width
-    height: Math.round(window.innerHeight * 0.75) + 'px',//height uses the window height as reference, the domNode did not worked
-
-    layout: {
-      hierarchical: {
-        enabled: true,
-        direction: "DU",//Down up orientation
-        sortMethod: 'directed',
-        levelSeparation: 120,//vertical space between the nodes
-        parentCentralization: true,
-        shakeTowards: "roots",//it moves the parents up in the pedigree
-        edgeMinimization: true,
-        blockShifting: true,
-        nodeSpacing: 140
+/**
+ * Returns `generations` from a pedigree of an individual given by an `id`
+ * string, using data from `genebanks`.
+ *
+ * @param genebanks genebank array to use for fetching individual data
+ * @param id individual number of the individual whose pedigree to generate
+ * @param generations the number of generations to plot
+ */
+export function calcPedigree(genebanks: Genebank[], id: string, generations: number = 1000): Pedigree {
+  let nodes: Node[] = []
+  let edges: Edge[] = []
+  const indData = getIndividual(genebanks, id)
+  if (indData) {
+    nodes.push( asNode(indData, edges.length) )
+
+    const getPedigree = (ind: Individual, level = generations) => {
+      let pedigreeNodes: Node[] = []
+      let pedigreeEdges: Edge[] = []
+      if (level > 0) {
+        // loop over mother and father, if they exist
+        [ind.mother, ind.father].forEach((parent: LimitedIndividual | null) => {
+          if (parent && parent.id !== null) {
+            const data = getIndividual(genebanks, parent.number)
+            if (data) {
+              pedigreeNodes.push(asNode(data, 0))
+              pedigreeEdges.push({id: `${ind.number}-${data.number}`,
+                                  from: ind.number,
+                                  to: data.number})
+              const parents = getPedigree(data, level - 1)
+              pedigreeNodes = [...pedigreeNodes, ...parents.nodes]
+              pedigreeEdges = [...pedigreeEdges, ...parents.edges]
+            }
+          }
+        })
       }
-    },
+      return {nodes: pedigreeNodes, edges: pedigreeEdges}
+    }
+    const pedigree = getPedigree(indData)
+    nodes = [...nodes, ...pedigree.nodes]
+    edges = [...edges, ...pedigree.edges]
+  }
+  // remove duplicate nodes and edges
+  nodes = nodes.filter((v,i,s) => s.findIndex(o => o.id == v.id) == i)
+  edges = edges.filter((v,i,s) => s.findIndex(o => o.id == v.id) == i)
```

Comment:
Creative solution! I have this in encam: 

```typescript
/** Returns a copy of the array with duplicates removed, via reference equality

  uniq([8,3,8,1]) // => [8,3,1]

  const x = {}
  const y = {}
  uniq([x,x,y,x,y]) // => [x,y]

*/
export function uniq<A>(xs: A[]): A[] {
  const seen = new Set()
  return xs.filter(x => {
    const duplicate = seen.has(x)
    seen.add(x)
    return !duplicate
  })
}
```

You can use it if you want, it's complexity is better (not quadratic). You might want to add a `by`/`key` argument so you can write `nodes = uniq(nodes, v => v.id)`

Reply:
hmm. yeah that's probably for the best :) I tend to write these chain-expressions while designing, but I need to learn to go back and clean them up! :D